packages/eslint-plugin/src/rules/no-invalid-this.ts

```diff
@@ -0,0 +1,159 @@
+import baseRule from 'eslint/lib/rules/no-invalid-this';
+import astUtils from 'eslint/lib/rules/utils/ast-utils';
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+import * as util from '../util';
+
+type CheckingContextNode =
+  | TSESTree.Program
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+interface CheckingContext {
+  init: boolean;
+  node: CheckingContextNode;
+  valid: boolean;
+}
+
+type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-invalid-this',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow `this` keywords outside of classes or class-like objects',
+      category: 'Best Practices',
+      recommended: false,
```

Comment:
Mark the rule as an extension rule.
You'll have to regen the configs and docs after this change.

```suggestion
      recommended: false,
      extendsBaseRule: true,
```